Disable EPOLLOUT if triggered while txbuf is empty

lnyng opened the pull request:
### Summary of the PR

Currently the vsock connection always triggers on EPOLLOUT. When a host applicationlistens to the UDS, it's almost always writable, and thus keeps triggering backend events with EPOLLOUT on the connection. As a result, vhost-device-vsock daemon consumes 100% CPU.

To make it more CPU efficient, it can instead unsubscribe EPOLLOUT whenever there is no more data to send. It can re-subscribe if the connection cannot write out all bytes. This change causes the CPU util to drop to negligible level.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

stefano-garzarella commented on vhost-device-vsock/src/vhu_vsock_thread.rs:
```diff
@@ -430,11 +430,19 @@ impl VhostUserVsockThread {
                 let conn = self.thread_backend.conn_map.get_mut(key).unwrap();
 
                 if evset.bits() == epoll::Events::EPOLLOUT.bits() {
+                    // If no remaining data to flush, try to disable EPOLLIN and then return
```
Typo in the comment?
IIUC we are disabling EPOLLOUT, right?

stefano-garzarella commented on vhost-device-vsock/src/vhu_vsock_thread.rs:
```diff
@@ -430,11 +430,19 @@ impl VhostUserVsockThread {
 
                 if evset.bits() == epoll::Events::EPOLLOUT.bits() {
+                    // If no remaining data to flush, try to disable EPOLLIN and then return
+                    if conn.tx_buf.is_empty() {
```
Is it okay to skip `self.thread_backend.backend_rxq.push_back()` call?

stefano-garzarella commented on vhost-device-vsock/src/vhu_vsock_thread.rs:
```diff
@@ -430,11 +430,19 @@ impl VhostUserVsockThread {
+                    // If no remaining data to flush, try to disable EPOLLIN and then return
+                    if conn.tx_buf.is_empty() {
+                        if Self::epoll_modify(epoll_fd, fd, epoll::Events::EPOLLIN).is_err() {
+                            warn!("epoll_modify failed");
```
we are emitting a warning here....

stefano-garzarella commented on vhost-device-vsock/src/vsock_conn.rs:
```diff
@@ -345,6 +345,16 @@ impl<S: AsRawFd + ReadVolatile + Write + WriteVolatile + IsHybridVsock> VsockCon
+            )
+            .is_err()
+            {
+                error!("epoll_modify failed");
```
....and an error here, can we do the same thing on both? 

And also put a better message, like in this case "Unable to re-enable EPOLLOUT" or something like that.

stefano-garzarella reviewed (commented):
Thanks for the PR and for the fix!

Where you able to bisect to understand if we had this issue from the beginning or we added it recently?

Please, can you also add a line in the changelog as specified in the PR template?

lnyng commented on vhost-device-vsock/src/vhu_vsock_thread.rs:
```diff
@@ -430,11 +430,19 @@ impl VhostUserVsockThread {
                     }
                 }
             } else {
+                let epoll_fd = self.get_epoll_fd();
```
> let conn = self.thread_backend.conn_map.get_mut(key).unwrap();

This line takes mut ref of self, so we can't call `self.get_epoll_fd()` after that line.

stefano-garzarella replied:
oh right, got it, but please can you move after the comment already there that is referred to the else branch or move that comment.

Something like this:
```diff
--- a/vhost-device-vsock/src/vhu_vsock_thread.rs
+++ b/vhost-device-vsock/src/vhu_vsock_thread.rs
@@ -428,10 +428,9 @@ impl VhostUserVsockThread {
                         .unwrap();
                     }
                 }
-            } else {
+            } else { // Previously connected connection
                 // Get epoll fd before getting conn as that takes self mut ref
                 let epoll_fd = self.get_epoll_fd();
-                // Previously connected connection
                 let key = self.thread_backend.listener_map.get(&fd).unwrap();
                 let conn = self.thread_backend.conn_map.get_mut(key).unwrap();
```

lnyng commented:
Thanks for the quick review! Addressed comments. I tried v0.1.0 but for some reason no connection can be established. I tried a few commits between v0.1.0 and v0.2.0 and all have the issue. This commit is the oldest I tried: https://github.com/rust-vmm/vhost-device/commit/6dabb28122301d0dafad606f5b80ba4b3edea461. Therefore it does not seem like a recent change.

stefano-garzarella commented on vhost-device-vsock/src/vhu_vsock_thread.rs:
```diff
@@ -18,7 +18,6 @@ use std::{
 };
 
-#[cfg(feature = "backend_vsock")]
 use log::error;
 use log::warn;
```
```suggestion
use log::{warn, error};
```

stefano-garzarella commented on vhost-device-vsock/src/vhu_vsock_thread.rs:
```diff
@@ -441,6 +442,13 @@ impl VhostUserVsockThread {
                             if cnt > 0 {
                                 conn.fwd_cnt += Wrapping(cnt as u32);
                                 conn.rx_queue.enqueue(RxOps::CreditUpdate);
+                            } else if conn.tx_buf.is_empty() {
```
Can we just check if `cnt == 0`?

IIUC `flush_to` returns `Ok(0)` if `tx_buf` is empty,

stefano-garzarella commented:
> Thanks for the quick review! Addressed comments. I tried v0.1.0 but for some reason no connection can be established. I tried a few commits between v0.1.0 and v0.2.0 and all have the issue. This commit is the oldest I tried: [6dabb28](https://github.com/rust-vmm/vhost-device/commit/6dabb28122301d0dafad606f5b80ba4b3edea461). Therefore it does not seem like a recent change.

Got it, thanks for checking!

lnyng commented:
@stsquad @epilys @vireshk Hi gentle ping for review request. Thanks!
